x-pack/plugins/security_solution/server/lib/detection_engine/rule_management/logic/detection_rules_client/methods/import_rule_with_source.ts

```diff
@@ -0,0 +1,102 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import type { RulesClient } from '@kbn/alerting-plugin/server';
+import type { ActionsClient } from '@kbn/actions-plugin/server';
+import { ruleTypeMappings } from '@kbn/securitysolution-rules';
+
+import type { RuleResponse } from '../../../../../../../common/api/detection_engine/model/rule_schema';
+import type { MlAuthz } from '../../../../../machine_learning/authz';
+import type { IPrebuiltRuleAssetsClient } from '../../../../prebuilt_rules/logic/rule_assets/prebuilt_rule_assets_client';
+import { createBulkErrorObject } from '../../../../routes/utils';
+import { convertAlertingRuleToRuleResponse } from '../converters/convert_alerting_rule_to_rule_response';
+import { convertRuleResponseToAlertingRule } from '../converters/convert_rule_response_to_alerting_rule';
+import type { ImportRuleArgs } from '../detection_rules_client_interface';
+import { applyRuleUpdate } from '../mergers/apply_rule_update';
+import { applyRuleDefaults } from '../mergers/apply_rule_defaults';
+import { validateMlAuth, toggleRuleEnabledOnUpdate } from '../utils';
+import { getRuleByRuleId } from './get_rule_by_rule_id';
+import { SERVER_APP_ID } from '../../../../../../../common';
+
+interface ImportRuleOptions {
+  actionsClient: ActionsClient;
+  rulesClient: RulesClient;
+  prebuiltRuleAssetClient: IPrebuiltRuleAssetsClient;
+  importRulePayload: ImportRuleArgs;
+  mlAuthz: MlAuthz;
+}
+
+/**
+ * Imports a rule with pre-calculated `rule_source` (and `immutable`) fields.
+ * Once prebuilt rule customization epic is complete, this function can be used
+ * to import all rules. Until then, this function should only be used when the feature flag is enabled.
+ *
+ */
+export const importRuleWithSource = async ({
+  actionsClient,
+  rulesClient,
+  importRulePayload,
+  prebuiltRuleAssetClient,
+  mlAuthz,
+}: ImportRuleOptions): Promise<RuleResponse> => {
+  const { ruleToImport, overwriteRules, allowMissingConnectorSecrets } = importRulePayload;
+  const { rule_source: ruleSource, immutable } = ruleToImport;
+
+  await validateMlAuth(mlAuthz, ruleToImport.type);
+
+  const existingRule = await getRuleByRuleId({
+    rulesClient,
+    ruleId: ruleToImport.rule_id,
+  });
+
+  if (existingRule && !overwriteRules) {
+    throw createBulkErrorObject({
+      ruleId: existingRule.rule_id,
+      statusCode: 409,
+      message: `rule_id: "${existingRule.rule_id}" already exists`,
+    });
```

Comment:
The Detection Rules Client shouldn't throw bulk error objects, as this is specific to route handlers. We need to throw a generic conflict error here.

Reply:
https://github.com/elastic/kibana/pull/190198/commits/5c5e6371ef4fc1737068c0af131a1f212dbc60ca

Reply:
P.S. I tried to remove any route/status_code responsibility from the client, and that ended up looking like [this](https://github.com/elastic/kibana/pull/190198/commits/1053da5408a9b8bed3b4433c051568b16cab7542), where we still pass a "conflict" error back up, and then the caller converts that to the `status_code: 409`. 

We need to pass that information back up _somehow_, and this seemed the cleanest without explicitly passing a status code. However, I noticed that the client is also sometimes throwing a [`ClientError`](https://github.com/elastic/kibana/blob/main/x-pack/plugins/security_solution/server/lib/detection_engine/rule_management/logic/detection_rules_client/methods/update_rule.ts#L50), which does contain a statusCode, so I'd like to be consistent one way or the other. @banderror @xcrzx any opinion here?

Reply:
Throwing errors with a status code looks fine to me. While it may not be 100% clean from an abstract architectural standpoint, it avoids introducing another intermediate error representation and having to convert it to a status code. That said, if you decide to move forward with an intermediate representation, that's also fine.

My main concern was about coupling the bulk error object, which is specific to API responses, with the import method, as it might be used in many other different contexts.